// @copyleaks/plagiarism-report/esm2020/lib/plagiarism-report/services/layout-media-query.service.mjs
import { Injectable } from '@angular/core';
import { distinctUntilChanged, map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@angular/flex-layout";
export class LayoutMediaQueryService {
    constructor(mediaObserver) {
        this.mediaObserver = mediaObserver;
        this.isActive = this.mediaObserver.isActive;
    }
    get isMobile$() {
        return this.mediaObserver.asObservable().pipe(map(changes => changes.map(ch => ch.mqAlias)), map(queries => queries.includes('xs') || queries.includes('sm')), distinctUntilChanged());
    }
    get mediaAliases$() {
        return this.mediaObserver.asObservable().pipe(map(changes => changes.map(ch => ch.mqAlias)));
    }
}
LayoutMediaQueryService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: LayoutMediaQueryService, deps: [{ token: i1.MediaObserver }], target: i0.ɵɵFactoryTarget.Injectable });
LayoutMediaQueryService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: LayoutMediaQueryService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: LayoutMediaQueryService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: i1.MediaObserver }]; } });
