// @copyleaks/plagiarism-report/esm2020/lib/plagiarism-report/services/view-mode.service.mjs
import { Injectable } from '@angular/core';
import { BehaviorSubject, Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { ALERTS } from '../utils/constants';
import * as i0 from "@angular/core";
import * as i1 from "./report.service";
export var EReportViewModel;
(function (EReportViewModel) {
    EReportViewModel[EReportViewModel["ScanningResult"] = 0] = "ScanningResult";
    EReportViewModel[EReportViewModel["Alerts"] = 1] = "Alerts";
    EReportViewModel[EReportViewModel["AIView"] = 2] = "AIView";
})(EReportViewModel || (EReportViewModel = {}));
export class ViewModeService {
    constructor(reportService) {
        this.reportService = reportService;
        this._reportViewMode$ = new BehaviorSubject(EReportViewModel.ScanningResult);
        this._unsub = new Subject();
        this.reportService.completeResult$.pipe(takeUntil(this._unsub)).subscribe(completeResult => {
            if (this.reportService.isPlagiarismEnabled()) {
                const hasAlerts = completeResult?.notifications?.alerts?.filter(s => s.code != ALERTS.SUSPECTED_AI_TEXT_DETECTED)?.length;
                const hasResults = !!completeResult.results?.batch?.length ||
                    !!completeResult.results?.internet?.length ||
                    !!completeResult.results?.database?.length ||
                    !!completeResult.results?.repositories?.length;
                if (hasAlerts && !hasResults) {
                    this._reportViewMode$.next(EReportViewModel.Alerts);
                }
            }
            else if (this.reportService.isAiDetectionEnabled()) {
                this.showAIAlertView(completeResult);
            }
        });
    }
    get reportViewMode$() {
        return this._reportViewMode$;
    }
    get selectedAlert() {
        return this._selectedAlert;
    }
    set selectedAlert(alert) {
        this._selectedAlert = alert;
    }
    showAIAlertView(completeResult) {
        this.reportService.configure({ contentMode: 'text' });
        const aiAlert = completeResult?.notifications?.alerts?.find(s => s.code == ALERTS.SUSPECTED_AI_TEXT_DETECTED);
        if (aiAlert) {
            this._selectedAlert = aiAlert;
        }
        this._reportViewMode$.next(EReportViewModel.AIView);
    }
    /**
     * change the view of the report
     * @param mode view mode
     */
    changeViewMode$(mode) {
        this._reportViewMode$.next(mode);
    }
    /**
     * unsubscribe from observalbes on destroy
     */
    ngOnDestroy() {
        this._reportViewMode$.unsubscribe();
    }
}
ViewModeService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: ViewModeService, deps: [{ token: i1.ReportService }], target: i0.ɵɵFactoryTarget.Injectable });
ViewModeService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: ViewModeService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: ViewModeService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.ReportService }]; } });
