// @copyleaks/plagiarism-report/esm2020/lib/plagiarism-report/components/exclude-partial-scan/exclude-partial-scan.component.mjs
import { Component, HostListener, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../../services/report.service";
export class ExcludePartialScanComponent {
    constructor(element, reportService) {
        this.element = element;
        this.reportService = reportService;
    }
    /**
     * emits the upgrade plan event
     */
    click() {
        this.reportService.upgradePlanEvent();
    }
}
ExcludePartialScanComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: ExcludePartialScanComponent, deps: [{ token: i0.ElementRef }, { token: i1.ReportService }], target: i0.ɵɵFactoryTarget.Component });
ExcludePartialScanComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.3", type: ExcludePartialScanComponent, selector: "span[cr-exclude-partial-scan]", inputs: { match: ["cr-exclude-partial-scan", "match"] }, host: { listeners: { "click": "click()" } }, ngImport: i0, template: '<ng-content></ng-content>', isInline: true, styles: [":host{filter:blur(3px);cursor:pointer}\n"] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: ExcludePartialScanComponent, decorators: [{
            type: Component,
            args: [{ selector: 'span[cr-exclude-partial-scan]', template: '<ng-content></ng-content>', styles: [":host{filter:blur(3px);cursor:pointer}\n"] }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i1.ReportService }]; }, propDecorators: { match: [{
                type: Input,
                args: ['cr-exclude-partial-scan']
            }], click: [{
                type: HostListener,
                args: ['click']
            }] } });
