// @copyleaks/plagiarism-report/esm2020/lib/plagiarism-report/models/api-models/CompleteResult.mjs
export var CompleteResultNotificationAlertSeverity;
(function (CompleteResultNotificationAlertSeverity) {
    CompleteResultNotificationAlertSeverity[CompleteResultNotificationAlertSeverity["VeryLow"] = 0] = "VeryLow";
    CompleteResultNotificationAlertSeverity[CompleteResultNotificationAlertSeverity["Low"] = 1] = "Low";
    CompleteResultNotificationAlertSeverity[CompleteResultNotificationAlertSeverity["Medium"] = 2] = "Medium";
    CompleteResultNotificationAlertSeverity[CompleteResultNotificationAlertSeverity["High"] = 3] = "High";
    CompleteResultNotificationAlertSeverity[CompleteResultNotificationAlertSeverity["VeryHigh"] = 4] = "VeryHigh";
})(CompleteResultNotificationAlertSeverity || (CompleteResultNotificationAlertSeverity = {}));
/** Enum representing the status of a scan */
export var ScanStatus;
(function (ScanStatus) {
    ScanStatus[ScanStatus["Success"] = 0] = "Success";
    ScanStatus[ScanStatus["Error"] = 1] = "Error";
    ScanStatus[ScanStatus["CreditsChecked"] = 2] = "CreditsChecked";
    ScanStatus[ScanStatus["Indexed"] = 3] = "Indexed";
})(ScanStatus || (ScanStatus = {}));
/** result preview types  */
export var EResultPreviewType;
(function (EResultPreviewType) {
    EResultPreviewType[EResultPreviewType["Batch"] = 0] = "Batch";
    EResultPreviewType[EResultPreviewType["Repositroy"] = 1] = "Repositroy";
    EResultPreviewType[EResultPreviewType["Internet"] = 2] = "Internet";
    EResultPreviewType[EResultPreviewType["Database"] = 3] = "Database";
})(EResultPreviewType || (EResultPreviewType = {}));
/** Enum representing the access of a result */
export var ResultAccess;
(function (ResultAccess) {
    ResultAccess[ResultAccess["full"] = 0] = "full";
    ResultAccess[ResultAccess["locked"] = 1] = "locked";
})(ResultAccess || (ResultAccess = {}));
