// @copyleaks/plagiarism-report/esm2020/lib/panel/panel.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MatButtonModule } from '@angular/material/button';
import { MatDividerModule } from '@angular/material/divider';
import { MatIconModule } from '@angular/material/icon';
import { FlexLayoutModule } from '@angular/flex-layout';
import { PanelComponent, PanelActionsComponent, PanelBodyComponent, PanelHeaderComponent, PanelActionComponent, } from './panel.component';
import * as i0 from "@angular/core";
export class PanelModule {
}
PanelModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: PanelModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PanelModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: PanelModule, declarations: [PanelComponent, PanelHeaderComponent, PanelActionsComponent, PanelActionComponent, PanelBodyComponent], imports: [CommonModule, FlexLayoutModule, MatIconModule, MatDividerModule, MatButtonModule], exports: [PanelComponent, PanelActionsComponent, PanelActionComponent, PanelHeaderComponent, PanelBodyComponent] });
PanelModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: PanelModule, imports: [[CommonModule, FlexLayoutModule, MatIconModule, MatDividerModule, MatButtonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: PanelModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [PanelComponent, PanelHeaderComponent, PanelActionsComponent, PanelActionComponent, PanelBodyComponent],
                    imports: [CommonModule, FlexLayoutModule, MatIconModule, MatDividerModule, MatButtonModule],
                    exports: [PanelComponent, PanelActionsComponent, PanelActionComponent, PanelHeaderComponent, PanelBodyComponent],
                }]
        }] });
