// @copyleaks/plagiarism-report/esm2020/lib/shared/pipes/safe.pipe.mjs
import { Pipe } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/platform-browser";
export class SafePipe {
    constructor(sanitizer) {
        this.sanitizer = sanitizer;
    }
    /**
     * A pipe to bypass the security trust for some `content` of some `type`
     * @param content the `content` to declare safe
     * @param type the the `type` of content
     */
    transform(content, type) {
        switch (type) {
            case 'html':
                return this.sanitizer.bypassSecurityTrustHtml(content);
            case 'style':
                return this.sanitizer.bypassSecurityTrustStyle(content);
            case 'script':
                return this.sanitizer.bypassSecurityTrustScript(content);
            case 'url':
                return this.sanitizer.bypassSecurityTrustUrl(content);
            case 'resourceUrl':
                return this.sanitizer.bypassSecurityTrustResourceUrl(content);
            default:
                throw new Error(`Invalid safe type specified: ${type}`);
        }
    }
}
SafePipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: SafePipe, deps: [{ token: i1.DomSanitizer }], target: i0.ɵɵFactoryTarget.Pipe });
SafePipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: SafePipe, name: "safe" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: SafePipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'safe',
                }]
        }], ctorParameters: function () { return [{ type: i1.DomSanitizer }]; } });
