// @copyleaks/plagiarism-report/esm2020/lib/shared/pipes/similarity.pipe.mjs
import { Pipe } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
export class SimilarityPipe {
    constructor(percentPipe) {
        this.percentPipe = percentPipe;
    }
    /**
     * Works like @angular/common PercentPipe but when value is larger than 0
     * and the result string is 0% , it returns < 1%.
     * See `PercentPipe` for params info.
     */
    transform(value, digitsInfo, locale) {
        let result = this.percentPipe.transform(value, digitsInfo, locale);
        if (value > 0 && parseFloat(result) === 0) {
            result = '< 1%';
        }
        if (value > 1) {
            result = '100%';
        }
        return result;
    }
}
SimilarityPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: SimilarityPipe, deps: [{ token: i1.PercentPipe }], target: i0.ɵɵFactoryTarget.Pipe });
SimilarityPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: SimilarityPipe, name: "similarity" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: SimilarityPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'similarity',
                }]
        }], ctorParameters: function () { return [{ type: i1.PercentPipe }]; } });
