// @copyleaks/plagiarism-report/esm2020/lib/plagiarism-report/components/html-helpers/suspect-html-helper.component.mjs
import { Component } from '@angular/core';
import { filter, map, withLatestFrom } from 'rxjs/operators';
import { untilDestroy } from '../../../shared/operators/untilDestroy';
import { MatchType } from '../../models';
import { findRespectiveStart } from '../../utils/match-helpers';
import { truthy } from '../../utils/operators';
import { HtmlHelperBase } from './HtmlHelperBase';
import iframeScript from './one-to-one-iframe-logic';
import * as i0 from "@angular/core";
import * as i1 from "../../services/highlight.service";
import * as i2 from "../../services/match.service";
import * as i3 from "../../services/report.service";
import * as i4 from "../../services/copyleaks-translate.service";
export class SuspectHtmlHelperComponent extends HtmlHelperBase {
    constructor(highlightService, matchService, renderer, element, reportService, translateService) {
        super(renderer, element, reportService, translateService);
        this.highlightService = highlightService;
        this.matchService = matchService;
        const js = renderer.createElement('script');
        js.textContent = iframeScript;
        this.script = js.outerHTML;
    }
    /**
     * Handle `match-select` event
     * @param event the event object
     */
    handleMatchSelect(event) {
        this.highlightService.setSuspectHtmlMatch(event.index !== -1 ? this.matches[event.index] : null);
    }
    /**
     * Life-cycle method
     * empty for `untilDestroy` rxjs operator
     */
    ngOnDestroy() { }
    /**
     * Life-cycle method
     * subscribe to:
     * - suspect changes
     * - view mode changes
     * - source and suspect html matches
     */
    ngOnInit() {
        const { suspectResult$: suspect$, contentMode$ } = this.reportService;
        const { sourceHtml$, textMatchClick$ } = this.highlightService;
        const { suspectHtmlMatches$ } = this.matchService;
        suspect$
            .pipe(untilDestroy(this), truthy())
            .subscribe(suspect => suspect.result && suspect.result.html && (this.html = suspect.result.html.value));
        suspectHtmlMatches$.pipe(untilDestroy(this)).subscribe(matches => this.renderMatches(matches));
        // jump$
        // 	.pipe(
        // 		untilDestroy(this),
        // 		withLatestFrom(viewMode$, contentMode$),
        // 		filter(([, view, content]) => view === 'one-to-one' && content === 'html')
        // 	)
        // 	.subscribe(([forward]) => this.messageFrame({ type: 'match-jump', forward } as MatchJumpEvent));
        textMatchClick$
            .pipe(untilDestroy(this), filter(ev => ev.origin === 'source' && ev.broadcast), map(ev => ev.elem), withLatestFrom(suspect$, suspectHtmlMatches$, contentMode$), filter(([, , matches, content]) => content === 'html' && !!matches))
            .subscribe(([elem, suspect, matches]) => {
            if (elem && suspect && suspect.result) {
                const comparison = suspect.result.html.comparison[MatchType[elem.match.type]];
                const [start] = findRespectiveStart(elem.match.start, comparison, true);
                const found = matches.findIndex(m => m.start === start);
                this.markSingleMatchInFrame(found);
            }
            else {
                this.markSingleMatchInFrame(-1);
            }
        });
        sourceHtml$
            .pipe(withLatestFrom(suspect$, suspectHtmlMatches$, contentMode$), filter(([, , matches, content]) => content === 'html' && !!matches), filter(([, suspect]) => suspect && suspect.result && !!suspect.result.html.value))
            .subscribe(([match, suspect, matches]) => {
            if (match && suspect && suspect.result) {
                const comparison = suspect.result.html.comparison[MatchType[match.type]];
                const [start] = findRespectiveStart(match.start, comparison, true);
                const found = matches.findIndex(m => m.start === start);
                this.markSingleMatchInFrame(found);
            }
            else {
                this.markSingleMatchInFrame(-1);
            }
        });
    }
}
SuspectHtmlHelperComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: SuspectHtmlHelperComponent, deps: [{ token: i1.HighlightService }, { token: i2.MatchService }, { token: i0.Renderer2 }, { token: i0.ElementRef }, { token: i3.ReportService }, { token: i4.CopyleaksTranslateService }], target: i0.ɵɵFactoryTarget.Component });
SuspectHtmlHelperComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.3", type: SuspectHtmlHelperComponent, selector: "iframe[cr-suspect-html-helper]", usesInheritance: true, ngImport: i0, template: '', isInline: true, styles: ["#sidebar.opened+#page-container{left:0px!important}#page-container .c{pointer-events:none!important}#page-container .t{pointer-events:all!important}span[exclude]{font-style:italic;opacity:.75}span[exclude-partial-scan]{filter:blur(6px);cursor:pointer}span[match]{cursor:pointer;color:#000}span[match] a:link{pointer-events:none}span[ignored]{font-style:italic}span[match][data-type=\"0\"]{background-color:#f66}span[match][data-type=\"0\"].hover{background-color:#ff8080}span[match][data-type=\"1\"]{background-color:#ff9a9a}span[match][data-type=\"1\"].hover{background-color:#ffb4b4}span[match][data-type=\"2\"]{background-color:#ffd9b0}span[match][data-type=\"2\"].hover{background-color:#ffe5ca}span[match][on]{background-color:gold!important}\n"] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: SuspectHtmlHelperComponent, decorators: [{
            type: Component,
            args: [{ selector: 'iframe[cr-suspect-html-helper]', template: '', styles: ["#sidebar.opened+#page-container{left:0px!important}#page-container .c{pointer-events:none!important}#page-container .t{pointer-events:all!important}span[exclude]{font-style:italic;opacity:.75}span[exclude-partial-scan]{filter:blur(6px);cursor:pointer}span[match]{cursor:pointer;color:#000}span[match] a:link{pointer-events:none}span[ignored]{font-style:italic}span[match][data-type=\"0\"]{background-color:#f66}span[match][data-type=\"0\"].hover{background-color:#ff8080}span[match][data-type=\"1\"]{background-color:#ff9a9a}span[match][data-type=\"1\"].hover{background-color:#ffb4b4}span[match][data-type=\"2\"]{background-color:#ffd9b0}span[match][data-type=\"2\"].hover{background-color:#ffe5ca}span[match][on]{background-color:gold!important}\n"] }]
        }], ctorParameters: function () { return [{ type: i1.HighlightService }, { type: i2.MatchService }, { type: i0.Renderer2 }, { type: i0.ElementRef }, { type: i3.ReportService }, { type: i4.CopyleaksTranslateService }]; } });
