// @copyleaks/plagiarism-report/esm2020/lib/mat-pagination/mat-pagination.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';
import { MatInputModule } from '@angular/material/input';
import { MatTooltipModule } from '@angular/material/tooltip';
import { FlexLayoutModule } from '@angular/flex-layout';
import { MatPaginationComponent } from './mat-pagination.component';
import { FormsModule } from '@angular/forms';
import * as i0 from "@angular/core";
export class MatPaginationModule {
}
MatPaginationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: MatPaginationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MatPaginationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: MatPaginationModule, declarations: [MatPaginationComponent], imports: [CommonModule,
        FormsModule,
        MatIconModule,
        FlexLayoutModule,
        MatButtonModule,
        MatInputModule,
        MatTooltipModule], exports: [MatPaginationComponent] });
MatPaginationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: MatPaginationModule, imports: [[
            CommonModule,
            FormsModule,
            MatIconModule,
            FlexLayoutModule,
            MatButtonModule,
            MatInputModule,
            MatTooltipModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: MatPaginationModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [MatPaginationComponent],
                    imports: [
                        CommonModule,
                        FormsModule,
                        MatIconModule,
                        FlexLayoutModule,
                        MatButtonModule,
                        MatInputModule,
                        MatTooltipModule,
                    ],
                    exports: [MatPaginationComponent],
                }]
        }] });
