// @copyleaks/plagiarism-report/esm2020/lib/plagiarism-report/components/match/match.component.mjs
import { Component, HostBinding, HostListener, Input } from '@angular/core';
import scrollIntoView from 'scroll-into-view-if-needed';
import * as i0 from "@angular/core";
import * as i1 from "../../services/highlight.service";
export class MatchComponent {
    constructor(element, renderer, highlightService) {
        this.element = element;
        this.renderer = renderer;
        this.highlightService = highlightService;
        this._focused = false;
    }
    /** focused flag, if set to `true` the element will be highlighted */
    get focused() {
        return this._focused;
    }
    set focused(value) {
        this._focused = value;
        if (value) {
            scrollIntoView(this.element.nativeElement, {
                behavior: 'smooth',
                block: 'nearest',
                skipOverflowHiddenElements: true,
            });
        }
    }
    /**
     * emits the element through the match click event
     */
    click() {
        this.highlightService.textMatchClicked({ elem: this, broadcast: true, origin: this.origin });
    }
    /**
     * Life-cycle method
     * Add a class to the native element that represents the match type
     */
    ngAfterContentInit() {
        this.renderer.addClass(this.element.nativeElement, `cr-m-${this.match.type}`);
    }
}
MatchComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: MatchComponent, deps: [{ token: i0.ElementRef }, { token: i0.Renderer2 }, { token: i1.HighlightService }], target: i0.ɵɵFactoryTarget.Component });
MatchComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.3", type: MatchComponent, selector: "span[cr-match]", inputs: { match: ["cr-match", "match"], origin: "origin" }, host: { listeners: { "click": "click()" }, properties: { "class.cr-highlight": "this.focused" } }, ngImport: i0, template: '<ng-content></ng-content>', isInline: true, styles: [".sat-tooltip{color:#fff;border-radius:4px;margin:14px;max-width:250px;padding:.5em;overflow:hidden;text-overflow:ellipsis;background:rgba(0,0,0,.8)}.identical{background-color:#ff796e}.minor-changes{background-color:#ff9a9a}.related{background-color:#ffd9b0}.omitted{background-color:#fff}.text-identical{color:#ff796e}.text-minor-changes{color:#ff9a9a}.text-related{color:#ffd9b0}.dot{border:1px solid #171717;width:1.5em;min-width:1.5em;min-height:1.5em;height:1.5em;border-radius:50%}@media screen and (max-width: 959px){.dot{width:1em;min-width:1em;min-height:1em;height:1em}}.copyleaks-report-dialog .mat-dialog-container{padding:0}.cdk-keyboard-focused{outline:2px solid black!important}:host{cursor:pointer;background-color:red}:host.cr-m-0{background-color:#ff796e}:host.cr-m-0:hover{background-color:#ff9188}span[class*=cr-m-]{cursor:pointer}span[class*=cr-m-].cr-m-1{background-color:#ff9a9a}span[class*=cr-m-].cr-m-1:hover{background-color:#ffb4b4}span[class*=cr-m-].cr-m-2{background-color:#ffd9b0}span[class*=cr-m-].cr-m-2:hover{background-color:#ffe5ca}span[class*=cr-m-].cr-m-100{color:#ffd9b0;text-decoration:underline;font-weight:bolder}span[class*=cr-m-].cr-highlight{background-color:gold}span[class*=cr-m-].cr-highlight:hover{background-color:#ffdf33}\n"] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: MatchComponent, decorators: [{
            type: Component,
            args: [{ selector: 'span[cr-match]', template: '<ng-content></ng-content>', styles: [".sat-tooltip{color:#fff;border-radius:4px;margin:14px;max-width:250px;padding:.5em;overflow:hidden;text-overflow:ellipsis;background:rgba(0,0,0,.8)}.identical{background-color:#ff796e}.minor-changes{background-color:#ff9a9a}.related{background-color:#ffd9b0}.omitted{background-color:#fff}.text-identical{color:#ff796e}.text-minor-changes{color:#ff9a9a}.text-related{color:#ffd9b0}.dot{border:1px solid #171717;width:1.5em;min-width:1.5em;min-height:1.5em;height:1.5em;border-radius:50%}@media screen and (max-width: 959px){.dot{width:1em;min-width:1em;min-height:1em;height:1em}}.copyleaks-report-dialog .mat-dialog-container{padding:0}.cdk-keyboard-focused{outline:2px solid black!important}:host{cursor:pointer;background-color:red}:host.cr-m-0{background-color:#ff796e}:host.cr-m-0:hover{background-color:#ff9188}span[class*=cr-m-]{cursor:pointer}span[class*=cr-m-].cr-m-1{background-color:#ff9a9a}span[class*=cr-m-].cr-m-1:hover{background-color:#ffb4b4}span[class*=cr-m-].cr-m-2{background-color:#ffd9b0}span[class*=cr-m-].cr-m-2:hover{background-color:#ffe5ca}span[class*=cr-m-].cr-m-100{color:#ffd9b0;text-decoration:underline;font-weight:bolder}span[class*=cr-m-].cr-highlight{background-color:gold}span[class*=cr-m-].cr-highlight:hover{background-color:#ffdf33}\n"] }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.Renderer2 }, { type: i1.HighlightService }]; }, propDecorators: { match: [{
                type: Input,
                args: ['cr-match']
            }], origin: [{
                type: Input
            }], focused: [{
                type: HostBinding,
                args: ['class.cr-highlight']
            }], click: [{
                type: HostListener,
                args: ['click']
            }] } });
