// @copyleaks/plagiarism-report/esm2020/lib/expansion-panel/expansion-panel.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ExpansionPanelComponent, ExpansionPanelTitleComponent, ExpansionPanelMenuComponent, ExpansionPanelMenuItemComponent, ExpansionPanelBodyComponent, } from './expansion-panel.component';
import { FlexLayoutModule } from '@angular/flex-layout';
import { MatButtonModule } from '@angular/material/button';
import { MatRippleModule } from '@angular/material/core';
import { MatDividerModule } from '@angular/material/divider';
import { MatIconModule } from '@angular/material/icon';
import { MatTooltipModule } from '@angular/material/tooltip';
import * as i0 from "@angular/core";
export class ExpansionPanelModule {
}
ExpansionPanelModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: ExpansionPanelModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ExpansionPanelModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: ExpansionPanelModule, declarations: [ExpansionPanelComponent,
        ExpansionPanelTitleComponent,
        ExpansionPanelMenuComponent,
        ExpansionPanelMenuItemComponent,
        ExpansionPanelBodyComponent], imports: [CommonModule,
        MatDividerModule,
        FlexLayoutModule,
        MatRippleModule,
        MatButtonModule,
        MatIconModule,
        MatTooltipModule], exports: [ExpansionPanelComponent,
        ExpansionPanelTitleComponent,
        ExpansionPanelMenuComponent,
        ExpansionPanelMenuItemComponent,
        ExpansionPanelBodyComponent] });
ExpansionPanelModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: ExpansionPanelModule, imports: [[
            CommonModule,
            MatDividerModule,
            FlexLayoutModule,
            MatRippleModule,
            MatButtonModule,
            MatIconModule,
            MatTooltipModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: ExpansionPanelModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        ExpansionPanelComponent,
                        ExpansionPanelTitleComponent,
                        ExpansionPanelMenuComponent,
                        ExpansionPanelMenuItemComponent,
                        ExpansionPanelBodyComponent,
                    ],
                    imports: [
                        CommonModule,
                        MatDividerModule,
                        FlexLayoutModule,
                        MatRippleModule,
                        MatButtonModule,
                        MatIconModule,
                        MatTooltipModule,
                    ],
                    exports: [
                        ExpansionPanelComponent,
                        ExpansionPanelTitleComponent,
                        ExpansionPanelMenuComponent,
                        ExpansionPanelMenuItemComponent,
                        ExpansionPanelBodyComponent,
                    ],
                }]
        }] });
