// @copyleaks/plagiarism-report/esm2020/lib/shared/shared.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule, PercentPipe } from '@angular/common';
import { ShortNumberPipe } from './pipes/short-number.pipe';
import { SafePipe } from './pipes/safe.pipe';
import { SearchPipe } from './pipes/search.pipe';
import { ArrayPipe } from './pipes/array.pipe';
import { ClickStopPropagationDirective } from './directives/click-stop-propagation.directive';
import { TimeAgoPipe } from './pipes/time-ago.pipe';
import { WhitelistPipe } from './pipes/whitelist.pipe';
import { UniquePipe } from './pipes/unique.pipe';
import { SimilarityPipe } from './pipes/similarity.pipe';
import * as i0 from "@angular/core";
/**
 * A shared module containing all sort of generic reuseable stuff
 */
export class SharedModule {
}
SharedModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: SharedModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SharedModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: SharedModule, declarations: [ShortNumberPipe,
        SafePipe,
        SearchPipe,
        ArrayPipe,
        TimeAgoPipe,
        WhitelistPipe,
        UniquePipe,
        SimilarityPipe,
        ClickStopPropagationDirective], imports: [CommonModule], exports: [ShortNumberPipe,
        SafePipe,
        SearchPipe,
        ArrayPipe,
        TimeAgoPipe,
        WhitelistPipe,
        SimilarityPipe,
        ClickStopPropagationDirective] });
SharedModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: SharedModule, providers: [PercentPipe], imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.3", ngImport: i0, type: SharedModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        ShortNumberPipe,
                        SafePipe,
                        SearchPipe,
                        ArrayPipe,
                        TimeAgoPipe,
                        WhitelistPipe,
                        UniquePipe,
                        SimilarityPipe,
                        ClickStopPropagationDirective,
                    ],
                    imports: [CommonModule],
                    providers: [PercentPipe],
                    exports: [
                        ShortNumberPipe,
                        SafePipe,
                        SearchPipe,
                        ArrayPipe,
                        TimeAgoPipe,
                        WhitelistPipe,
                        SimilarityPipe,
                        ClickStopPropagationDirective,
                    ],
                }]
        }] });
